File: misc/gauss_filter_model.m
clc; close all;
TSW = 7;
cic1 = ones(2 ^ TSW, 1);
cic2 = conv(cic1, cic1);
cic4 = conv(cic2, cic2);
cic4 = cic4 / sum(cic4);

a = 28621495321396;
b = 29171251135283;

order = 32;
scale = 2 ^ order;

p = [a * ones(1000, 1); b * ones(1000, 1)];
size(p)
size(cic4)

am = floor(a / scale);
bm = floor(b / scale);

pm = [am * ones(1000, 1); bm * ones(1000, 1)];
size(pm);

pf = conv(cic4, p);
size(pf)

pfm = conv(cic4, pm) * scale;
%figure(1, WindowStyle="docked")
figure('windowstyle','docked')
    clf
    hold on
        plot(p)
        plot(pf)
        plot(pfm)
    hold off
    grid

File: misc/gauss_filter_model.py
import numpy as np
import matplotlib.pyplot as plt

# Константы и параметры
TSW = 7
order = 32
a = 28621495321396
b = 29171251135283
scale = 2 ** order

# Создание фильтра CIC4
cic1 = np.ones(2 ** TSW)
cic2 = np.convolve(cic1, cic1)
cic4 = np.convolve(cic2, cic2)
cic4 = cic4 / np.sum(cic4)

# Формирование сигнала
p = np.concatenate((a * np.ones(1000), b * np.ones(1000)))

# Масштабированный сигнал
am = a // scale
bm = b // scale
pm = np.concatenate((am * np.ones(1000), bm * np.ones(1000)))

# Применение фильтра CIC4 к сигналу
pf = np.convolve(cic4, p)
pfm = np.convolve(cic4, pm) * scale

# Визуализация результатов
plt.figure()
plt.plot(p, label="Исходный сигнал")
plt.plot(pf, label="Сглаженный сигнал (pf)")
plt.plot(pfm, label="Сглаженный масштабированный сигнал (pfm)")
plt.legend()
plt.grid()
plt.show()
